### PR TITLE
FlySky IBUS receiver support

### PR DESCRIPTION
**Pull request by sidhantgoel**
FlySky IBUS receiver support. RX needs to be connected to DSM port.

### CODE REVIEW

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        return false;
+    }
+    if ((frame[1] != 0x40)) {
+        return false;
+    }
+    uint16_t checksum = (frame[31] << 8) | frame[30];
+    uint16_t calcsum = 0xFFFF;
+    for(i = 0; i < 30; i++) calcsum -= frame[i];
+    if(checksum != calcsum) {
+        uint16_t values1[IBUS_INPUT_CHANNELS] = {1000, 1400, 1000, 1000, 1000, 1000};
```
Comment (WickedShell) [inline libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: Momentary corruption killing all the channels seems like a poor choice, as well as the fact that this defeats some of the debounce internally, and you don't actually know what channel the user has mapped mode/throttle onto for example.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        add_input(6, values1, false);
+       return false;
+    }
+
+    *nvalues = max_values;
+    if(*nvalues > 14) *nvalues = 14;
```
Comment (WickedShell) [inline libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: Please put this on a seperate line and use `{ }` as per the style guide.

Comment (WickedShell) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: This is also clearer to be written as `*nvalues = MIN(max_values, 14)` rather then the awkward pointer dereference/if.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    byte_input.buf[byte_input.ofs++] = b;
+
+    if (byte_input.ofs == sizeof(byte_input.buf)) {
+        uint16_t values[IBUS_INPUT_CHANNELS];
+        uint16_t num_values=IBUS_INPUT_CHANNELS;
+        bool ibus_failsafe = false;
+        //bool sbus_frame_drop = false;
+        if(ibus_decode(byte_input.buf, values, &num_values, IBUS_INPUT_CHANNELS)) {
+            add_input(num_values, values, ibus_failsafe);
```
Comment (WickedShell) [inline libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: This is always false for `ibus_failsafe`. Remove the variable and just pass `false` in here for clarity.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    //     return;
+    // }
+
+    byte_input.buf[byte_input.ofs++] = b;
+
+    if (byte_input.ofs == sizeof(byte_input.buf)) {
+        uint16_t values[IBUS_INPUT_CHANNELS];
+        uint16_t num_values=IBUS_INPUT_CHANNELS;
+        bool ibus_failsafe = false;
+        //bool sbus_frame_drop = false;
```
Comment (WickedShell) [inline libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: This is dead/old, no need to carry it forward, just delete it.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        // definately not IBUS, missing header byte
+        return;
+    }
+    
+    if (byte_input.ofs == 1 && b != 0x40) {
+        // definately not IBUS, missing header byte
+        byte_input.ofs = 0;
+        return;
+    }
+    // if (byte_input.ofs == 32 && byte_input.buf[1] == 0x20 && byte_input.buf[2] == 0x40) {
```
Comment (WickedShell) [inline libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: This is dead/old, no need to carry it forward, just delete it.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    /* check frame boundary markers to avoid out-of-sync cases */
+    if ((frame[0] != 0x20)) {
+        return false;
+    }
+    if ((frame[1] != 0x40)) {
+        return false;
+    }
+    uint16_t checksum = (frame[31] << 8) | frame[30];
+    uint16_t calcsum = 0xFFFF;
+    for(i = 0; i < 30; i++) calcsum -= frame[i];
```
Comment (WickedShell) [inline libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: Use a new line, `{ }`'s around the loop. This should also probably be using `IBUS_FRAME_SIZE - 2` rather then the magic 30. You carefully defined it but never used it.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    //     #endif
+    if (ss.process_pulse(w0, w1, b)) {
+        //uint16_t values[6] = {1000, 1000, 1000, 1000, 1000, 1000};
+        //add_input(6, values, false);
+        _process_byte(b);
+    }
+}
+
+// decode a full IBUS frame
+bool AP_RCProtocol_IBUS::ibus_decode(const uint8_t frame[32], uint16_t *values, uint16_t *nvalues, uint16_t max_values)
```
[the remark this answers is not in the document]
Comment (WickedShell) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: This should also be using the define rather then the magic 32.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 2 above)
Comment (WickedShell) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: Also use your define, rather then the magic value of 14.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 1 above)
Comment (sidhantgoel) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: Sorry, that code I was using for testing, will remove it.

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (sidhantgoel) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: It is not passing on stack, the number is only useful for sizeof, will use define for packet size

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 2 above)
Comment (sidhantgoel) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: ok

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 3 above)
Comment (sidhantgoel) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: ok

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 4 above)
Comment (sidhantgoel) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: ok

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 5 above)
Comment (sidhantgoel) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: ok

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 6 above)
Comment (sidhantgoel) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: ok

File: libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (sidhantgoel) [reply libraries/AP_RCProtocol/AP_RCProtocol_IBUS.cpp]: oops my bad, subscript is ignored by the compiler, even sizeof returns sizeof pointer

**Comment by sidhantgoel** [comment]
@WickedShell Hey, I've done all the requested changes, how will this merge to master.
@tridge @bugobliterator

**Comment by PraiseSatan** [comment]
Hi, I had also been working on implementing ibus support into ardupilot and seem to have followed a similar approach.
If you leave in the soft serial processing of pulses, I don't see a reason to disable processing it as pulses, and that would also allow the reciever to be connected to the RC in port (which for me at least means I can actually plug it in as I don't have the required connector for the DSM port).

However more importantly, the ibus receiver I have been using has failsafe behaviour which your code doesn't account for.
With some channels set up with failsafe it uses the high value bits in channel data which would typically be unused. This allows some detection of failsafe and without properly accounting for it can cause a very high value to be sent.

Some more details are here:
https://github.com/ArduPilot/ardupilot/issues/9419#issuecomment-453903677

**Comment by tridge** [comment]
as this is a duplicate of #10240 I'm going to close this one
@sidhantgoel if you see anything in #10240 that doesn't look right then please comment there

**Comment by fiechr** [comment]
I just connected a Fli14+ receiver to a fmuv3 (ArduCopter V4.0.0-rc1 (ec01ae4c)) Pixhawk's RC IN connector and it works fine (10 channels available with FS-i6 and modified firmware). Maybe even 14 channels are available, but I don't have enough controls. ;)

I set the generic "RX Setup" on the FS-i6 remote to "I-BUS/PPM" (as far as I can tell it's always possible to set whatever you like at this option, but that doesn't mean it's really supported or the receiver is able to switch protocols at all).

I'm also using a Flysky iA6B and a Flysky iA10B on other Pixhawk devices.
AFAIK iA6B does not support S-BUS, only iA10B does. But to be honest, I'm not really sure which protocols are effectively getting used on which device anymore.

Is there a direct way to find out which protocol (PPM, S-BUS, I-BUS) is currently used on the running device? Is it possible to force a certain protocol if there is/seems more than one option available?

I also just wanted to offer my support for testing, debugging the I-BUS support, also since I have a few Flyksy receivers available at the moment.

**Comment by rmackay9** [comment]
@fiechr, it might be better if this was raised in the [Copter-4.0 discussion area](https://discuss.ardupilot.org/c/arducopter/copter-40) but in any case I think the protocol being used is written in the logs and should also be visible on Mission Planner's message tab as shown below.
mp-rcinput-msg

**Pull request closed without merging**